File: solutions/3160. Find the Number of Distinct Colors Among the Balls/3160.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
 public:
  vector<int> queryResults(int limit, const vector<vector<int>>& queries) {
    unordered_map<int, int> ballToColor;
    unordered_map<int, int> colorCount;
    vector<int> ans;

    for (const vector<int>& query : queries) {
      const int ball = query[0];
      const int color = query[1];
      if (const auto it = ballToColor.find(ball); it != ballToColor.cend()) {
        const int prevColor = it->second;
        if (--colorCount[prevColor] == 0)
          colorCount.erase(prevColor);
      }
      ballToColor[ball] = color;
      ++colorCount[color];
      ans.push_back(colorCount.size());
    }

    return ans;
  }
};

File: solutions/3160. Find the Number of Distinct Colors Among the Balls/3160.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
  def queryResults(self, limit: int, queries: List[List[int]]) -> List[int]:
    ballToColor = {}
    colorCount = collections.Counter()
    ans = []

    for ball, color in queries:
      if ball in ballToColor:
        prevColor = ballToColor[ball]
        colorCount[prevColor] -= 1
        if colorCount[prevColor] == 0:
          del colorCount[prevColor]
      ballToColor[ball] = color
      colorCount[color] += 1
      ans.append(len(colorCount))

    return ans